### PR TITLE
Optimized dirt and snow spreading

### PR DESCRIPTION
**Pull request by lukalt**
Significantly reduces the number of chunk lookups when grass is spread in random ticks. This does not change the game mechanics.

### CODE REVIEW

**Review by Titaniumtown — APPROVED** [review]
I've been testing this patch on my server for a while now w/o any issues. LGTM.

**Review by MiniDigger — APPROVED** [review]

**Review by imDaniX — CHANGES_REQUESTED** [review]
I need some _space_ plz
`if (expression)`

File: Spigot-Server-Patches/0666-optimized-dirt-and-snow-spreading.patch [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+-        return b(iblockdata, iworldreader, blockposition) && !iworldreader.getFluid(blockposition1).a((Tag) TagsFluid.WATER);
++        return canBeGrass(chunk, iblockdata, iworldreader, blockposition) && !chunk.getFluid(blockposition1).a((Tag) TagsFluid.WATER); // Paper - reuse chunk
+     }
+ 
+     @Override
+     public void tick(IBlockData iblockdata, WorldServer worldserver, BlockPosition blockposition, Random random) {
+         if (this instanceof BlockGrass && worldserver.paperConfig.grassUpdateRate != 1 && (worldserver.paperConfig.grassUpdateRate < 1 || (MinecraftServer.currentTick + blockposition.hashCode()) % worldserver.paperConfig.grassUpdateRate != 0)) { return; } // Paper
+-        if (!b(iblockdata, (IWorldReader) worldserver, blockposition)) {
++        // Paper - reuse chunk
++        Chunk chunk = worldserver.getChunkIfLoaded(blockposition);
```
Comment (Fox2Code) [inline Spigot-Server-Patches/0666-optimized-dirt-and-snow-spreading.patch]: I really like the fact that it avoids loading chunks for spreading grass.

**Review by Fox2Code — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Review by PaulBGD — APPROVED** [review]
Looks good, although admittedly this will only improve dirt/snow spreading by about 30%, as the light check is the heaviest part of spreading.

**Review by Owen1212055 — APPROVED** [review]

**Pull request merged**